cmd/config.go

```diff
@@ -274,12 +276,30 @@ type BeelineConfig struct {
 	Dataset string
 	// SampleRate is the (positive integer) denominator of the sample rate.
 	// Default: 1 (meaning all traces are sent). Set higher to send fewer traces.
-	SampleRate uint
+	SampleRate uint32
 	// Mute disables honeycomb entirely; useful in test environments.
 	Mute bool
 	// Many other fields of beeline.Config are omitted as they are not yet used.
 }
 
+// makeSampler constructs a SamplerHook which will deterministically decide if
+// any given span should be sampled based on its TraceID, which is shared by all
+// spans within a trace. A sample rate of 0 defaults to a sample rate of 1
+// (i.e. all events are sent).
+func makeSampler(rate uint32) func(fields map[string]interface{}) (bool, int) {
+	if rate == 0 {
+		rate = 1
+	}
+	upperBound := math.MaxUint32 / rate
+
+	return func(fields map[string]interface{}) (bool, int) {
+		h := sha1.Sum([]byte(fields["trace.trace_id"].(string)))
+		// Pack the first four bytes of the sha1 into a single uint32.
+		v := (uint32(h[0]) << 24) | (uint32(h[1]) << 16) | (uint32(h[2]) << 8) | uint32(h[3])
+		return v < upperBound, int(rate)
+	}
```

Comment:
We don't need this to be cryptographically secure so let's use the faster fnv, which conveniently has a way to emit a 32-bit number directly: https://golang.org/pkg/hash/fnv/#New32

Reply:
Done! I'm not a fan of the "hashers are io.Writers" API but this does seem more straightforward.